### ISSUE TITLE
styles are only loaded for first component test if styled-components is used

### ISSUE THREAD

**Issue by JochenDiekenbrock** [cypress-io/cypress#15879]
### Current behavior

I'm migrating React component tests from Cypress 6.8 to 7.0.1 and I followed the migration instructions. The tests are generally running. However there is a problem with styled-components. The <style></style> element for the styled based css is only present for the first test in a file (even if the it() tests are in different describe() blocks).
If I run just the second test by using it.only() then the styles are present in this test so I'm sure that it's not my testcase itself that's causing this problem.

### Desired behavior

The styles must be loaded in every test.

### Test code to reproduce

see https://github.com/JochenDiekenbrock/dzi-search
Running cypress open-ct will create two screen shots, both should be identical.

### Versions

Cypress 7.0.1
Linux

In a private repo, this problem did not occur with Cypress 6.8.0

**Comment by glomotion** [comment]
I too am seeing this with my react component testing. 
Very first mount of a component renders perfectly, the second mount of the same component renders without the injected styles.

We're using @emotion/css - which behaves quite similarly to styled-components. 

deps setup is as follows:
```
"@emotion/css": "^11.1.3",
"@emotion/babel-plugin": "^11.3.0",
"@cypress/code-coverage": "^3.8.2",
"@cypress/react": "^5.3.2",
"@cypress/webpack-dev-server": "^1.1.2",
"cypress": "^7.0.1",
```

**Comment by glomotion** [comment]
FWIW, one possible workaround I have discovered so far, is to split up the it('test') definitions between different files. This ensures that the component always renders with the correct injected styles. 

Not great, but it's a start... :P

**Comment by lmiller1990** [comment]
Hi @JochenDiekenbrock,

I cloned your repo but I got the same screenshots when I ran the tests:

![image](https://user-images.githubusercontent.com/19196536/114135002-f8298680-994b-11eb-913b-2924e7f94500.png)

Do I need to do something specific to reproduce this? Or @glomotion can you provide a repo (if you have one set up and it's not too much trouble - or you can just paste all the files here and I can use those to make it). Otherwise I can try to reproduce it.

If I had to guess where the problem might be, [somewhere around here](https://github.com/cypress-io/cypress/blob/726120da183cb6d89c83181761d21f09844a9fc2/npm/react/src/mount.ts#L155) where we clean up between tests.

**Comment by JochenDiekenbrock** [comment]
test-1.png and test-2.png should be exactly identical because they do exactly the same
```describe('SearchResults', () => {
  it('should render 1', () => {
    mountComponent({
      organizations: [{ title: 'Org 1' }, { title: 'Org 2' }]
    } as any);
    cy.screenshot('test-1');
  });

  it('should render 2', () => {
    mountComponent({
      organizations: [{ title: 'Org 1' }, { title: 'Org 2' }]
    } as any);
    cy.screenshot('test-2');
  });
});```

The error is, that `margin-bottom: 1rem;` is defined for each line. For test-1.png, which runs first, the margin is there, for test-2.png it's missing because the styled styles are not injected.

**Comment by glomotion** [comment]
@lmiller1990 my thoughts too. 
I remember reading that cypress 7.0.1 + cypress/react no longer persists dom modifications inbetween test suites. 
Sadly, im also seeing this in a private repo. Shouldnt be too hard to setup a demo tho. I might try and do so tonight.

**Comment by glomotion** [comment]
OK done! 
@lmiller1990 try checking this repo out and run it locally (`npm run test:cy:watch`)
https://github.com/glomotion/test-emotion-cypress-7-x

**Comment by glomotion** [comment]
both tests are identical. heres how each gets rendered
![Screen Shot 2021-04-09 at 4 40 20 pm](https://user-images.githubusercontent.com/1452237/114139378-50638700-9952-11eb-8441-2976189d8f73.png)

![Screen Shot 2021-04-09 at 4 40 12 pm](https://user-images.githubusercontent.com/1452237/114139364-4d689680-9952-11eb-90de-9fe60f97b0f5.png)

**Comment by glomotion** [comment]
oh man!! I thought for sure this would fix the issue: 
```
import { CacheProvider } from "@emotion/react";
import createCache from "@emotion/cache";

const cssCache = createCache({
  key: "mooo-cow",
  container: document.body, // eg put the styles into the body, instead of the head of the doc (which gets reset after each test)
  prepend: true,
});

describe("<AssetDetailsContent />", () => {
  it("test 1", () => {
    mount(
      <CacheProvider value={cssCache}>
        <Box backgroundColor={colors.bg[700]} padding="30px">
          <NotFoundPage />
        </Box>
      </CacheProvider>
    );
  });
});

...
```

Doesn't seem to help. The styles still get injected into the head, and then cleared out at the completion of the test run.

**Comment by glomotion** [comment]
@lmiller1990 i have some additional info on this. 

I tinkered with my tests a little, to have `@emotion/css` inject its styles into `document.body`, instead of the `document.head` (thinking this would short circuit the logic that clears out the doc head after each test run - such as the code reference you posted above). 

see this line: 
```
beforeEach(() => {
  sheet.container = document.body;
});
```

Sadly, this doesn't help - all styles are getting blown away after each test - even though they're not in the head of the document. See these 2 screenshots: 

<img width="1458" alt="Screen Shot 2021-04-11 at 9 18 37 am" src="https://user-images.githubusercontent.com/1452237/114287112-af86e000-9aa7-11eb-9563-e279023bec3e.png">

<img width="1456" alt="Screen Shot 2021-04-11 at 9 18 13 am" src="https://user-images.githubusercontent.com/1452237/114287110-ab5ac280-9aa7-11eb-8aa3-c17b4cbbb60d.png">

**Comment by glomotion** [comment]
OK, so i have found a hacky way to work around this issue - just for now. Emotion/css allows you to tap into the css creation/caching - and vary what happens with stylesheets. 

For now, i'm just duplicating all style sheets and re-injecting them before each test run: 
```
beforeEach(() => {
    sheet.tags.forEach((tag) => {
      document.body.appendChild(tag);
    });
  });
```
@JochenDiekenbrock, i'm pretty sure styled-components will allow a similar escape hatch - so this should probably work for you too. 

@lmiller1990 please sing out if i can be of any further help here, i'd definitely love to see this issue fixed somehow on the cypress side, though admittedly i'm not quite sure what can be done atm, other than reverting the decision to clear out document.head after each test run. 🤔

**Comment by davincho** [comment]
I had exactly the same issue - as far as I understood emotion (in my case v10) does not re-created the the required css class (in both runs it results into the same name).
My work around was creating a new emotion cache every test render:

```jsx

import createCache from '@emotion/cache';

<CacheProvider
      value={createCache({
        key: Math.random()
          .toString(36)
          .replace(/[^a-z]+/g, '')
          .substr(0, 5)
      })}
    >
  {children}
</CacheProvider>
```

**Comment by glomotion** [comment]
@lmiller1990 so i'm not sure, but i may have slammed into a related issue, this time with cy.intercepts. 

In short, when you create a bunch of `cy.intercept` stubs (inside of beforeEach) - they function perfectly for the very first test run, but for the second, third etc test runs they do absolutely nothing. 

here is an example intercept: 
```
cy.intercept(
  {
    method: 'GET',
    pathname: '/v1/collections',
  },
  req => {
    req.reply({
      fixture: 'mockCollections.json',
    });
  },
).as('collectionsStub');
```

It's almost as if something in the interactive component test UI isn't tearing down and or restoring cypress state/functionality inbetween test runs. 

I ran the exact same code (except with a `cy.visit('http://localhost:3000')` instead of a mount()) inside of the regular e2e test runner, and the intercepts function correctly for every single test run - this leads me to think that it's not a problem with the test code, and more likely a bug.

**Comment by lmiller1990** [comment]
Interesting, so the network layer may be involved. And it only happens in the component testing runner (not the e2e runner)?

What do you mean by "they function perfectly for the very first test run, but for the second, third etc test runs they do absolutely nothing". The requests are not intercepted, and just carry on as usual (presumably returning a 404?)

**Comment by glomotion** [comment]
> And it only happens in the component testing runner (not the e2e runner)?

Correct! very weird! :P

> The requests are not intercepted, and just carry on as usual (presumably returning a 404?)

Correct! The intercepts do absolutely nothing and the requests get made as normal (in the 2nd, 3rd, 4th tests). When running the first test, the intercepts correctly kick in and stub the API call, and return fixture data.

**Comment by glomotion** [comment]
OK so turns out `cy.intercept` works fine, when these calls are made simply using something like axios. 
see the revised test code here: https://github.com/glomotion/test-emotion-cypress-7-x/blob/master/src/NotFoundPage.cy.test.tsx

In the (private) repo where i'm seeing this `cy.intercept` weirdness, all our api calls are orchestrated by Redux Sagas - which fire off axios.get() requests inside of generator functions.

It's going to take me a while longer to add redux and saga's to the repo demo above. 
I'll try to do that at some point soon. 

For now sadly, my team have had to put plans to migrate from RTL to `cypress/react` on the back burner until it's a bit more stable. 😭

**Comment by glomotion** [comment]
Oddly enough though, the `cy.intercept` is still allowing the API call through, which is strange. 🤔

**Comment by JochenDiekenbrock** [comment]
I just updated my project (https://github.com/JochenDiekenbrock/dzi-search) to Cypress 7.1.0 and gave it a try. The error is still there.

**Comment by gilamran** [comment]
**Hope this helps someone / solving this issue:**

We encountered the same issue when using `makeStyles` in material-ui.
Under the hood it uses `jss`, we found a workaround by providing `jss` an insertion point, like this:

```
import React from 'react';
import { jssPreset, StylesProvider, ThemeProvider } from '@material-ui/core/styles';
import { create } from 'jss';

const styleNode = document.createComment('jss-insertion-point');
document.head.insertBefore(styleNode, document.head.firstChild);

const jss = create({
  ...jssPreset(),
  insertionPoint: 'jss-insertion-point',
});

export const App: React.FC = () => (
  <StylesProvider injectFirst jss={jss}>
    <ThemeProvider theme={theme}>
      <Main />
    </ThemeProvider>
  </StylesProvider>
);

```

**Comment by glomotion** [comment]
@lmiller1990 I think i have found the culprit for our cy.intercept() woes. 
It's an API library which we're importing / working with which uses `fp-ts` - amount other things. 
I'll share more info as I uncover it, but for now, after creating the minimal repro repo, I cant repro the issue.

**Comment by lmiller1990** [comment]
@glomotion cool, glad to hear `cy.intercept` is working correctly and the problem is not here - this method has been battle tested for years, but regressions can always happen so it's good to scrutinize, especially after a major release.

What is blocking your migration? It is just the styled components issue (sounds like `cy.intercept` is something in your code base).

@gilamran nice workaround, I'll try it out too - probably be good to at least have this documented, or possibly include something in `cypress/react` to make this less painful.

**Comment by lmiller1990** [comment]
Mini reproduction here: https://github.com/cypress-io/cypress/pull/15997

I will push my work and explorations here, and find a solution. I'm not exactly sure the best way to handle this yet.

1. some special logic for injected styles/scripts that should *not* be cleaned up
2. some way to tell cypress not to clean certain elements
3. something else?
4. better docs

**Comment by lmiller1990** [comment]
Ok, I think we have a few options.

## Just don't clean up `<head>`

This is pretty similar to what Jest will do. If you modify `document.head`, it will be persisted between tests. Generally, cleaning up between tests is good, but this is an option.

I think we will still clean up `<body>` - this makes sense, since we want to mount a new component each test.

## Partial opt-out of cleaning up

We could provide some way to opt-out of specific clean up. In the example of `styled-components`, this is injected:

```
<style data-styled="active" data-styled-version="5.0.0">.qBWUh{margin-bottom:1rem;}</style>
```

A potential opt-out API might be something like this:

```js
const Line = styled.div`
  margin-bottom: 1rem
`

const SearchResults = (props) => (
  <Line>Stuff</Line>
)

mount(<SearchResults />, {
  persist: '[data-styled]'
})
```

Anything found via the `querySelector` syntax passed into `persist` could not be cleaned up. This would likely be confusing, though, since most people probably don't think about how `styled-components` work, and would have to make the connection before coming across the `persist` option.

## Full opt-out

```js
mount(<SearchResults />, {
  cleanHead: true
})
```

I guess we'd probably default to... `false`? Although it feels correct to clean up between tests, in practice it seems like most of the time this is not the desired behavior.

**Comment by glomotion** [comment]
> @glomotion cool, glad to hear `cy.intercept` is working correctly and the problem is not here - this method has been battle tested for years, but regressions can always happen so it's good to scrutinize, especially after a major release.
> 
> What is blocking your migration? It is just the styled components issue (sounds like `cy.intercept` is something in your code base).

yeah so ATM we're fine with the workaround that we have for emotion/injected styles. The real blocker for us is cy.intercept() - but i still don't know enough about the lib that we're utilising to muse as to what its doing that is tanking the cypress network layer. Im still investigating.

**Comment by JochenDiekenbrock** [comment]
> Ok, here are my proposals. There are several options.
> ## Just don't clean up `<head>`
...
> ## Partial opt-out of cleaning up
...
> ## Full opt-out

Let me start by mentioning that I have full trust that you guys will choose whatever option is best given your overall user feedback and experience. So for me any of the options is fine.

My feeling is that, similar to when a test pollutes global, if it pollutes head it's the responsibility of the test to cleanup head if required. The points to the "don't clean up" option.

The "cleanHead" option extends this thinking by offering a service to the user to do the cleanup, which is of course nice to have.

The partial option seems to be the most effort on Cypress side and at the same time requires most knowledge from the user. And it is relatively fragile because the selector might not work anymore after a third party lib changes their internals.

My personal ranking therefore would be

1. full opt out
2. don't clean up head
3. partial opt out

**Comment by JochenDiekenbrock** [comment]
The problem that I see with this approach is that (right now, don't have the time to check) I don't know how to manually inject the styled-component style. Together with the (relative) inaccessibility of the react-scripts based build configuration to get notified of the generated files, manually injecting might be difficult.

**Comment by Develliot** [comment]
I'm with @JochenDiekenbrock I don't have a clue how to work around this. I also don't understand why it works when you split tests up into different files but not when you split it in to different "it" statements. 

I think if the default behaviour is going to blow stuff away and rebuild it between tests then, then I think I should be done in the same way that it does when it loads a new test file. That or opt out of what ever is breaking this for styled components. It would be nice to set this option at a high level like the cypress config so you don't need to set this in every mount of every test.

**Comment by elevatebart** [comment]
@Develliot @JochenDiekenbrock I am going to try to explain why it is failing.

Pardon my English if it is not too clear, English is not my first language.

In the component testing runner, we load a spec file once.
We then execute each it statement within the file.
When the component loads, the loader installs the styles for the component in a place of its choosing. 
Since before each test we cleanup the installed components and re-init the header for everything that has beed added during the test. If we did it after each test, we would not see anything in the renderings.

Come to our issue here: The styles added by `styled-components` are added in the zone of the DOM that is NUKED before each test. We have to figure out a way to leave these styles alone somehow.

**Comment by Develliot** [comment]
Is the component loading with each mount? Should it not rebuild the stuff in the headers at this point in time?

**Comment by lmiller1990** [comment]
Yeah, while it's nice to know how/why things work, most people just want to tests their components  - I don't think people should need to understand how things like styled-components actually work to be able to test them.

Let's clean up what Cypress injects (we have a `styles` mounting option) but leave the userland things alone (such as injections from styled-components and friends).

I will make a PR today. Thanks for the input everyone!

**Comment by lmiller1990** [comment]
Ok, here is the PR: https://github.com/cypress-io/cypress/pull/15997

We will no longer clean up `<head>` or `<body>`, except for things injected by Cypress. For example, you can inject a stylesheet like so:

```js
mount(<Foo />, {
  stylesheets: ['https://boostrap.com/styles.css']
})
```

We *do* clean up that stylesheet - this is expected, since it's part of the `mount` API, and of course we *do* clean up the mounted component in between each test. We will *not* clean up stylesheets injected from userland libraries, like `styled-components`, Emotion, etc. This should generally match the behavior most users will expect.

This solves the use cases presented here. If any others arise, we can just patch them as we go, but I can't think of any edge cases off the top of my head.

**Comment by glomotion** [comment]
@lmiller1990 apologies i haven't been able to participate in the conversation the last 2 days or so. This all sounds wonderful. From the bottom of my team's heart, thank you for this! 😍

**Comment by JochenDiekenbrock** [comment]
@lmiller1990 Thanks a lot from me as well. Great job!

**Comment by Develliot** [comment]
Thanks everyone this all sounds excellent

**Comment by Shaegi** [comment]
I got the same problem. After some research you can use the styled-components  `StylesheetManager` and its `disableCSSOMInjection` prop which seems to bypass this issue. 

Setup would be something like this: 
```JSX
mount(<StyleSheetManager disableCSSOMInjection={true}>
  <ThemeProvider theme={theme}>
      <App />
  </ThemeProvider>
</StyleSheetManager>
)
```

"Switches to the slower text node-based CSS injection system for adding styles to the DOM. Useful for integrating with third party tools that haven't been upgraded to consume styles from the CSSOM APIs."
Further read: https://styled-components.com/docs/api#stylesheetmanager

Also another usefull blog about this: https://blog.logrocket.com/using-the-stylesheetmanager-component-in-styled-components-5-0/

**Comment by elevatebart** [comment]
This has been released in version 5.3.4 of `@cypress/react`.

Can you try and see if it works please?

If it does, you can close this issue ;)

**Comment by JochenDiekenbrock** [comment]
It works fine with https://github.com/JochenDiekenbrock/dzi-search after updating to 5.3.4. Thanks a lot!
